In [628]:
import numpy as np
import pandas as pd
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn import model_selection
import matplotlib.pyplot as plt
from sklearn.linear_model import RidgeCV, LassoCV

In [629]:
car = pd.read_csv("/Users/huangyixiang/Desktop/udemy/UNZIP_FOR_NOTEBOOKS_FINAL/MLHomework/assignment4.csv")

In [630]:
car.head(20)

Unnamed: 0   mpg  cylinders  displacement  horsepower  weight  \
0            0  18.0          8         307.0       130.0    3504   
1            1  15.0          8         350.0       165.0    3693   
2            2  18.0          8         318.0       150.0    3436   
3            3  16.0          8         304.0       150.0    3433   
4            4  17.0          8         302.0       140.0    3449   
5            5  15.0          8         429.0       198.0    4341   
6            6  14.0          8         454.0       220.0    4354   
7            7  14.0          8         440.0       215.0    4312   
8            8  14.0          8         455.0       225.0    4425   
9            9  15.0          8         390.0       190.0    3850   
10          10  15.0          8         383.0       170.0    3563   
11          11  14.0          8         340.0       160.0    3609   
12          12  15.0          8         400.0       150.0    3761   
13          13  14.0          8         455.0       225.0    3086   
14          14  24.0          4         113.0        95.0    2372   
15          15  22.0          6         198.0        95.0    2833   
16          16  18.0          6         199.0        97.0    2774   
17          17  21.0          6         200.0        85.0    2587   
18          18  27.0          4          97.0        88.0    2130   
19          19  26.0          4          97.0        46.0    1835   

    acceleration  model year  origin  
0           12.0          70       1  
1           11.5          70       1  
2           11.0          70       1  
3           12.0          70       1  
4           10.5          70       1  
5           10.0          70       1  
6            9.0          70       1  
7            8.5          70       1  
8           10.0          70       1  
9            8.5          70       1  
10          10.0          70       1  
11           8.0          70       1  
12           9.5          70       1  
13          10.0          70       1  
14          15.0          70       3  
15          15.5          70       1  
16          15.5          70       1  
17          16.0          70       1  
18          14.5          70       3  
19          20.5          70       2

In [631]:
car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    398 non-null    int64  
 1   mpg           398 non-null    float64
 2   cylinders     398 non-null    int64  
 3   displacement  398 non-null    float64
 4   horsepower    398 non-null    float64
 5   weight        398 non-null    int64  
 6   acceleration  398 non-null    float64
 7   model year    398 non-null    int64  
 8   origin        398 non-null    int64  
dtypes: float64(4), int64(5)
memory usage: 28.1 KB


In [632]:
car = car.drop(['Unnamed: 0'], axis = 1)

In [633]:
car.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model year      0
origin          0
dtype: int64

In [634]:
car['origin'] = car['origin'].map({1: 'USA', 2: 'Europe', 3: 'Japan'})
car = pd.get_dummies(car, columns=['origin'], prefix='', prefix_sep='')
car_dataset = car.drop(['USA'], axis = 1)

In [635]:
car_dataset.head(50)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0   18.0          8         307.0       130.0    3504          12.0   
1   15.0          8         350.0       165.0    3693          11.5   
2   18.0          8         318.0       150.0    3436          11.0   
3   16.0          8         304.0       150.0    3433          12.0   
4   17.0          8         302.0       140.0    3449          10.5   
5   15.0          8         429.0       198.0    4341          10.0   
6   14.0          8         454.0       220.0    4354           9.0   
7   14.0          8         440.0       215.0    4312           8.5   
8   14.0          8         455.0       225.0    4425          10.0   
9   15.0          8         390.0       190.0    3850           8.5   
10  15.0          8         383.0       170.0    3563          10.0   
11  14.0          8         340.0       160.0    3609           8.0   
12  15.0          8         400.0       150.0    3761           9.5   
13  14.0          8         455.0       225.0    3086          10.0   
14  24.0          4         113.0        95.0    2372          15.0   
15  22.0          6         198.0        95.0    2833          15.5   
16  18.0          6         199.0        97.0    2774          15.5   
17  21.0          6         200.0        85.0    2587          16.0   
18  27.0          4          97.0        88.0    2130          14.5   
19  26.0          4          97.0        46.0    1835          20.5   
20  25.0          4         110.0        87.0    2672          17.5   
21  24.0          4         107.0        90.0    2430          14.5   
22  25.0          4         104.0        95.0    2375          17.5   
23  26.0          4         121.0       113.0    2234          12.5   
24  21.0          6         199.0        90.0    2648          15.0   
25  10.0          8         360.0       215.0    4615          14.0   
26  10.0          8         307.0       200.0    4376          15.0   
27  11.0          8         318.0       210.0    4382          13.5   
28   9.0          8         304.0       193.0    4732          18.5   
29  27.0          4          97.0        88.0    2130          14.5   
30  28.0          4         140.0        90.0    2264          15.5   
31  25.0          4         113.0        95.0    2228          14.0   
32  25.0          4          98.0        93.5    2046          19.0   
33  19.0          6         232.0       100.0    2634          13.0   
34  16.0          6         225.0       105.0    3439          15.5   
35  17.0          6         250.0       100.0    3329          15.5   
36  19.0          6         250.0        88.0    3302          15.5   
37  18.0          6         232.0       100.0    3288          15.5   
38  14.0          8         350.0       165.0    4209          12.0   
39  14.0          8         400.0       175.0    4464          11.5   
40  14.0          8         351.0       153.0    4154          13.5   
41  14.0          8         318.0       150.0    4096          13.0   
42  12.0          8         383.0       180.0    4955          11.5   
43  13.0          8         400.0       170.0    4746          12.0   
44  13.0          8         400.0       175.0    5140          12.0   
45  18.0          6         258.0       110.0    2962          13.5   
46  22.0          4         140.0        72.0    2408          19.0   
47  19.0          6         250.0       100.0    3282          15.0   
48  18.0          6         250.0        88.0    3139          14.5   
49  23.0          4         122.0        86.0    2220          14.0   

    model year  Europe  Japan  
0           70       0      0  
1           70       0      0  
2           70       0      0  
3           70       0      0  
4           70       0      0  
5           70       0      0  
6           70       0      0  
7           70       0      0  
8           70       0      0  
9           70       0      0  
10          70       0      0  

# Question1

In [636]:
X = car_dataset.drop('mpg',axis=1)
y = car_dataset['mpg']

80% training and 20% test

In [637]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [638]:
lasso_model1 = Lasso(alpha=1)
lasso_model1.fit(X_train, y_train)

Lasso(alpha=1)

Lasso_model1

In [639]:
lasso_model1.coef_

array([ 0.        , -0.        , -0.00532359, -0.00665824,  0.        ,
        0.704171  ,  0.        ,  0.        ])

20% training and 80% test

In [640]:
X = car_dataset.drop('mpg',axis=1)
y = car_dataset['mpg']

In [641]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.80, random_state=42)

In [642]:
lasso_model2 = Lasso(alpha=1)
lasso_model2.fit(X_train, y_train)

Lasso(alpha=1)

Lasso_model2

In [643]:
lasso_model2.coef_

array([-0.        , -0.00562763, -0.02435025, -0.00598054,  0.        ,
        0.61970055,  0.        , -0.        ])

# Question2

For 80% training and 20 test

In [644]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [645]:
import time
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LassoLarsIC
from sklearn.pipeline import make_pipeline

start_time = time.time()
lasso_lars_ic = make_pipeline(
    StandardScaler(), LassoLarsIC(criterion="aic", normalize=False)
).fit(X_train, y_train)
fit_time = time.time() - start_time

In [646]:
results = pd.DataFrame(
    {
        "alphas": lasso_lars_ic[-1].alphas_,
        "AIC criterion": lasso_lars_ic[-1].criterion_,
    }
).set_index("alphas")
alpha_aic = lasso_lars_ic[-1].alpha_

In [647]:
lasso_lars_ic.set_params(lassolarsic__criterion="bic").fit(X_train, y_train)
results["BIC criterion"] = lasso_lars_ic[-1].criterion_
alpha_bic = lasso_lars_ic[-1].alpha_

In [648]:
def highlight_min(x):
    x_min = x.min()
    return ["font-weight: bold" if v == x_min else "" for v in x]


results.style.apply(highlight_min)

In [649]:
min_alphas = 0.06750691622651057

In [650]:
lasso_model_min = Lasso(alpha=0.06750691622651057)
lasso_model_min.fit(X_train, y_train)

Lasso(alpha=0.06750691622651057)

In [651]:
lasso_model_min.intercept_

-19.931249456996433

In [652]:
lasso_model_min.coef_

array([-0.        ,  0.01087387, -0.00875756, -0.00699346,  0.05274951,
        0.810288  ,  1.77189689,  1.52511945])

Explain how to choose the optimal alpha:
We start the alpha from 0 and fit the model for each alpha to check the AIC and BIC, with the alpha increasing, both AIC and BIC will reach their own lowest value, then find the coressponding alpha as the optimal point. In this case, we choose the alpha with lowest AIC


# Question3

In [653]:
X = car_dataset.drop('mpg',axis=1)
y = car_dataset['mpg']

In [654]:
model_3 = LinearRegression()

In [655]:
from sklearn.model_selection import KFold

In [656]:
kfold = KFold(n_splits=10, shuffle=True, random_state=42)

In [657]:
results = cross_val_score(model, X, y, cv=kfold, scoring='r2')

In [658]:
print(results)

[0.82644355 0.87831208 0.83368074 0.80213669 0.84758022 0.81205758
 0.77665039 0.80919246 0.77321497 0.79711136]


In [659]:
a=[0.82644355,0.87831208,0.83368074,0.80213669,0.84758022,0.81205758,
 0.77665039,0.80919246,0.77321497,0.79711136]

In [660]:
sum(a)/len(a)

0.8156380040000002

Observation: The highest R-square of the 10-fold cross validation is 0.84758022, the mean R-square is 0.815638, because we use the complete dataset to fit model, the R-square is not OOS R2, it just show how well the model fit the data

# Question4

In [661]:
X = car_dataset.drop('mpg',axis=1)
y = car_dataset['mpg']

In [662]:
log(398)

5.986452005284438

In [663]:
lasso_model1.coef_

array([ 0.        , -0.        , -0.00532359, -0.00665824,  0.        ,
        0.704171  ,  0.        ,  0.        ])

In [664]:
lasso_model2.coef_

array([-0.        , -0.00562763, -0.02435025, -0.00598054,  0.        ,
        0.61970055,  0.        , -0.        ])

In [665]:
lasso_model_min.coef_

array([-0.        ,  0.01087387, -0.00875756, -0.00699346,  0.05274951,
        0.810288  ,  1.77189689,  1.52511945])

In [666]:
yhat_1 = lasso_model1.predict(X)

In [667]:
yhat_2 = lasso_model2.predict(X)

In [668]:
yhat_min = lasso_model_min.predict(X)

In [669]:
from math import log
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
 
# calculate aic for regression
def calculate_aic(n, mse, num_params):
    aic = n * log(mse) + 2 * num_params
    return aic
 
# number of parameters
num_params = len(lasso_model1.coef_) + 1
print('Number of parameters: %d' % (num_params))
# predict the training set

# calculate the error
mse = mean_squared_error(y, yhat_1)
print('MSE: %.3f' % mse)
# calculate the aic
aic = calculate_aic(len(y), mse, num_params)
print('AIC: %.3f' % aic)


Number of parameters: 9
MSE: 11.745
AIC: 998.442


In [670]:
# calculate aic for regression
def calculate_aic(n, mse, num_params):
    aic = n * log(mse) + 2 * num_params
    return aic
 
# number of parameters
num_params = len(lasso_model2.coef_) + 1
print('Number of parameters: %d' % (num_params))
# predict the training set

# calculate the error
mse = mean_squared_error(y, yhat_2)
print('MSE: %.3f' % mse)
# calculate the aic
aic = calculate_aic(len(y), mse, num_params)
print('AIC: %.3f' % aic)

Number of parameters: 9
MSE: 12.541
AIC: 1024.536


In [671]:
def calculate_aic(n, mse, num_params):
    aic = n * log(mse) + 2 * num_params
    return aic
 
# number of parameters
num_params = len(lasso_model_min.coef_) + 1
print('Number of parameters: %d' % (num_params))

# calculate the error
mse = mean_squared_error(y, yhat_min)
print('MSE: %.3f' % mse)
# calculate the aic
aic = calculate_aic(len(y), mse, num_params)
print('AIC: %.3f' % aic)

Number of parameters: 9
MSE: 10.932
AIC: 969.882


In [672]:
print("Lasso_model1 AIC: 998.442")
print("Lasso_model2 AIC: 1024.536")
print("Lasso_model_min AIC: 969.882")

Lasso_model1 AIC: 998.442
Lasso_model2 AIC: 1024.536
Lasso_model_min AIC: 969.882


In [673]:
def calculate_bic(n, mse, num_params):
    bic = n * log(mse) + num_params * log(n)
    return bic


# number of parameters
num_params = len(lasso_model1.coef_) + 1
print('Number of parameters: %d' % (num_params))

# calculate the error
mse = mean_squared_error(y, yhat_1)
print('MSE: %.3f' % mse)
# calculate the bic
bic = calculate_bic(len(y), mse, num_params)
print('BIC: %.3f' % bic)

Number of parameters: 9
MSE: 11.745
BIC: 1034.320


In [674]:
def calculate_bic(n, mse, num_params):
    bic = n * log(mse) + num_params * log(n)
    return bic


# number of parameters
num_params = len(lasso_model2.coef_) + 1
print('Number of parameters: %d' % (num_params))

# calculate the error
mse = mean_squared_error(y, yhat_2)
print('MSE: %.3f' % mse)
# calculate the bic
bic = calculate_bic(len(y), mse, num_params)
print('BIC: %.3f' % bic)

Number of parameters: 9
MSE: 12.541
BIC: 1060.414


In [675]:
def calculate_bic(n, mse, num_params):
    bic = n * log(mse) + num_params * log(n)
    return bic


# number of parameters
num_params = len(lasso_model_min.coef_) + 1
print('Number of parameters: %d' % (num_params))

# calculate the error
mse = mean_squared_error(y, yhat_min)
print('MSE: %.3f' % mse)
# calculate the bic
bic = calculate_bic(len(y), mse, num_params)
print('BIC: %.3f' % bic)

Number of parameters: 9
MSE: 10.932
BIC: 1005.760


In [676]:
print("Lasso_model1 BIC: 1034.320")
print("Lasso_model2 BIC: 1060.414")
print("Lasso_model_min BIC: 1005.760")

Lasso_model1 BIC: 1034.320
Lasso_model2 BIC: 1060.414
Lasso_model_min BIC: 1005.760


In [677]:
AIC_1 = 998.442
AIC_2 = 1024.536
AIC_min = 969.882

In [678]:
n = len(car_dataset)
n

398

In [679]:
k = 9

In [680]:
AICc_1 = AIC_1 + (2 * k**2 + 2 * k)/(n-k-1)
AICc_1

998.9059175257732

In [681]:
AICc_2 = AIC_2 + (2 * k**2 + 2 * k)/(n-k-1)
AICc_2

1024.9999175257733

In [682]:
AICc_min = AIC_min + (2 * k**2 + 2 * k)/(n-k-1)
AICc_min

970.3459175257732

In [683]:
print("Lasso_model1 AICc: 998.906")
print("Lasso_model2 AICc: 1025")
print("Lasso_model_min AICc: 970.346")

Lasso_model1 AICc: 998.906
Lasso_model2 AICc: 1025
Lasso_model_min AICc: 970.346


In [684]:
print("Lasso_model1 AIC: 998.442")
print("Lasso_model2 AIC: 1024.536")
print("Lasso_model_min AIC: 969.882")

print("Lasso_model1 BIC: 1034.320")
print("Lasso_model2 BIC: 1060.414")
print("Lasso_model_min BIC: 1005.760")

print("Lasso_model1 AICc: 998.906")
print("Lasso_model2 AICc: 1025")
print("Lasso_model_min AICc: 970.346")



Lasso_model1 AIC: 998.442
Lasso_model2 AIC: 1024.536
Lasso_model_min AIC: 969.882
Lasso_model1 BIC: 1034.320
Lasso_model2 BIC: 1060.414
Lasso_model_min BIC: 1005.760
Lasso_model1 AICc: 998.906
Lasso_model2 AICc: 1025
Lasso_model_min AICc: 970.346


1.Lasso_model_min(from Question3) is the best model because it has lowes AIC, BIC and AICc
2.BIC is always larger than AIC, because AIC = deviance + 2df; BIC = deviance + log(n)df, in this data set, log(n) is larger than 2.

# Question5

What are Randomized Control Trials: A/B Tests? What is its significance? Elaborate. Explain methods used in analyzing the RCT data

RCT:
The Randomized Control Trials is a scientific design that completely randomizes samples into two groups, treatment group, and control group. Because the sampling and grouping are completely random, both known and unknown factors will be balanced or we can say eliminated, then we can just focus on the treatment itself to know whether the treatment is impactful or not.

Significance:
It's commonly used in the clinical trial or other medical experiments. Many tech companies use A/B test to test their hypothesis for any update or new product, to see if the new one is better than the old one. Because it's impossible to create two completely identical groups as comparison, RCT is the best way to test the treatment effect.

Method:
Treatment mean minus Control mean.